sys/net/gnrc/link_layer/gnrc_lorawan/gnrc_lorawan_mcps.c

```diff
@@ -0,0 +1,341 @@
+{
+    /* Encrypt payload (it's block encryption so we can use the same buffer!) */
+    gnrc_lorawan_encrypt_payload((iolist_t *) payload, &mac->dev_addr, mac->mcps.fcnt, GNRC_LORAWAN_DIR_UPLINK, port ? mac->appskey : mac->nwkskey);
+
+    /* We try to allocate the whole header with fopts at once */
+    uint8_t fopts_length = gnrc_lorawan_build_options(mac, NULL);
+
+    gnrc_pktsnip_t *mac_hdr = gnrc_pktbuf_add(payload, NULL, sizeof(lorawan_hdr_t) + fopts_length + 1, GNRC_NETTYPE_UNDEF);
+
+    if (!mac_hdr) {
+        gnrc_pktbuf_release_error(payload, -ENOBUFS);
+        return NULL;
+    }
+
+    gnrc_pktsnip_t *mic = gnrc_pktbuf_add(NULL, NULL, MIC_SIZE, GNRC_NETTYPE_UNDEF);
+    if (!mic) {
+        gnrc_pktbuf_release_error(mac_hdr, -ENOBUFS);
+        return NULL;
+    }
+
+    lorawan_buffer_t buf = {
+        .data = (uint8_t *) mac_hdr->data,
+        .size = mac_hdr->size,
+        .index = 0
+    };
+
+    gnrc_lorawan_build_hdr(confirmed_data ? MTYPE_CNF_UPLINK : MTYPE_UNCNF_UPLINK,
+                           &mac->dev_addr, mac->mcps.fcnt, mac->mcps.ack_requested, fopts_length, &buf);
+
+    gnrc_lorawan_build_options(mac, &buf);
+
+    assert(buf.index == mac_hdr->size - 1);
+
+    buf.data[buf.index++] = port;
+
+    gnrc_lorawan_calculate_mic(&mac->dev_addr, mac->mcps.fcnt, GNRC_LORAWAN_DIR_UPLINK,
+                               (iolist_t *) mac_hdr, mac->nwkskey, mic->data);
+
+    LL_APPEND(payload, mic);
+
+    return mac_hdr;
+}
+
+static void _end_of_tx(gnrc_lorawan_t *mac, int type, int status)
+{
+    mac->mcps.waiting_for_ack = false;
+
+    mcps_confirm_t *mcps_confirm = _mcps_allocate(mac);
+
+    mcps_confirm->type = type;
+    mcps_confirm->status = status;
+    mac->netdev.event_callback((netdev_t *) mac, NETDEV_EVENT_MCPS_CONFIRM);
+
+    mac->mcps.fcnt += 1;
+}
+
+void gnrc_lorawan_mcps_event(gnrc_lorawan_t *mac, int event, int data)
+{
+    if (mac->mlme.activation == MLME_ACTIVATION_NONE) {
+        return;
+    }
+
+    if (event == MCPS_EVENT_ACK_TIMEOUT) {
+        gnrc_lorawan_send_pkt(mac, mac->mcps.outgoing_pkt, mac->last_dr);
+    }
+    else {
+        int state = mac->mcps.waiting_for_ack ? MCPS_CONFIRMED : MCPS_UNCONFIRMED;
+        if (state == MCPS_CONFIRMED && ((event == MCPS_EVENT_RX && !data) ||
+                                        event == MCPS_EVENT_NO_RX)) {
+            if (mac->mcps.nb_trials-- == 0) {
+                _end_of_tx(mac, MCPS_CONFIRMED, -ETIMEDOUT);
+            }
+        }
+        else {
+            _end_of_tx(mac, state, GNRC_LORAWAN_REQ_STATUS_SUCCESS);
+        }
+
+        mac->msg.type = MSG_TYPE_MCPS_ACK_TIMEOUT;
+        if (mac->mcps.outgoing_pkt) {
+            xtimer_set_msg(&mac->rx, 1000000 + random_uint32_range(0, 2000000), &mac->msg, thread_getpid());
+        }
+    }
+}
+
+void gnrc_lorawan_mcps_request(gnrc_lorawan_t *mac, mcps_request_t *mcps_request, mcps_confirm_t *mcps_confirm)
```

Comment:
to me it looks like you can omit the `mcps_confirm_t` param and make this function `int` (instead of `void`) and return the status code instead of using `mcps_confirm->status`. Because the only usage of this function is [here](https://github.com/RIOT-OS/RIOT/pull/11022/files#diff-385d2a03ea8a97a25dbb4c48d127beb1R210) where you pass in an empty `mcps_confirm_t` just to use its status variable for return afterwards.

Reply:
the problem with that is MCPS and MLME confirm can return other attributes (besides status, Link Check stats, MIB params, etc). The current implementation still doesn't add too many MCPS and MLME confirmations, so it's not that visible yet

Reply:
are you OK with this statement?